File: app/src/main/java/net/ballmerlabs/uscatterbrain/network/wifidirect/WifiDirectRadioModule.java
package net.ballmerlabs.uscatterbrain.network.wifidirect;

import android.net.wifi.p2p.WifiP2pInfo;

import net.ballmerlabs.uscatterbrain.db.ScatterbrainDatastore;
import net.ballmerlabs.uscatterbrain.db.entities.HashlessScatterMessage;
import net.ballmerlabs.uscatterbrain.db.entities.ScatterMessage;
import net.ballmerlabs.uscatterbrain.network.BlockHeaderPacket;
import net.ballmerlabs.uscatterbrain.network.BlockSequencePacket;
import net.ballmerlabs.uscatterbrain.network.bluetoothLE.BootstrapRequest;

import io.reactivex.Completable;
import io.reactivex.Flowable;
import io.reactivex.Single;
import io.reactivex.subjects.CompletableSubject;

public interface WifiDirectRadioModule {
    String TAG = "WifiDirectRadioModule";
    Single<WifiP2pInfo> connectToGroup(String name, String passphrase, int timeout);
    Completable bootstrapFromUpgrade(BootstrapRequest upgradeRequest);
    void unregisterReceiver();
    void registerReceiver();

    class BlockDataStream {
        private final Flowable<BlockSequencePacket> sequencePackets;
        private final BlockHeaderPacket headerPacket;
        private final ScatterMessage messageEntity;
        private final CompletableSubject sequenceCompletable = CompletableSubject.create();

        public BlockDataStream(BlockHeaderPacket headerPacket, Flowable<BlockSequencePacket> sequencePackets) {
            this.sequencePackets = sequencePackets
                    .doOnComplete(sequenceCompletable::onComplete)
                    .doOnError(sequenceCompletable::onError);
            this.headerPacket = headerPacket;
            if (this.headerPacket == null) {
                throw new IllegalStateException("header packet was null");
            }
            messageEntity = new ScatterMessage();
            messageEntity.message = new HashlessScatterMessage();
            messageEntity.message.to = headerPacket.getToFingerprint().toByteArray();
            messageEntity.message.from = headerPacket.getFromFingerprint().toByteArray();
            messageEntity.message.application = headerPacket.getApplication();
            messageEntity.message.sig = headerPacket.getSignature();
            messageEntity.message.sessionid = headerPacket.getSessionID();
            messageEntity.message.blocksize = headerPacket.getBlockSize();
            messageEntity.message.mimeType =  headerPacket.getMime();
            messageEntity.message.extension = headerPacket.getExtension();
            messageEntity.message.globalhash = ScatterbrainDatastore.getGlobalHash(headerPacket.getHashList());
            messageEntity.messageHashes = HashlessScatterMessage.hash2hashs(headerPacket.getHashList());
        }

        public Completable await() {
            return sequencePackets.ignoreElements();
        }

        public BlockDataStream(ScatterMessage message, Flowable<BlockSequencePacket> packetFlowable) {
            this(message, packetFlowable, false, true);
        }

        public boolean getToDisk() {
            return headerPacket.getToDisk();
        }

        public BlockDataStream(ScatterMessage message, Flowable<BlockSequencePacket> packetFlowable, boolean end, boolean todisk) {
            BlockHeaderPacket.Builder builder = BlockHeaderPacket.newBuilder()
                    .setToFingerprint(message.message.to)
                    .setFromFingerprint(message.message.from)
                    .setApplication(message.message.application)
                    .setSig(message.message.sig)
                    .setToDisk(todisk)
                    .setSessionID(message.message.sessionid)
                    .setBlockSize(message.message.blocksize)
                    .setMime(message.message.mimeType)
                    .setExtension(message.message.extension)
                    .setHashes(HashlessScatterMessage.hashes2hash(message.messageHashes));

            if (end) {
                builder.setEndOfStream();
            }

            this.headerPacket = builder.build();
            this.messageEntity = message;
            this.sequencePackets = packetFlowable
                    .doOnComplete(sequenceCompletable::onComplete)
                    .doOnError(sequenceCompletable::onError);
            if (this.headerPacket == null) {
                throw new IllegalStateException("header packet was null");
            }
        }

        public Completable awaitSequencePackets() {
            return sequenceCompletable;
        }

        public BlockHeaderPacket getHeaderPacket() {
            return headerPacket;
        }

        public Flowable<BlockSequencePacket> getSequencePackets() {
            return sequencePackets;
        }

        public ScatterMessage getEntity() {
            return messageEntity;
        }
    }
}

File: app/src/main/java/net/ballmerlabs/uscatterbrain/network/wifidirect/WifiDirectRadioModule.kt
package net.ballmerlabs.uscatterbrain.network.wifidirect

import io.reactivex.Completable
import io.reactivex.Flowable
import io.reactivex.Observable
import io.reactivex.Single
import io.reactivex.subjects.CompletableSubject
import net.ballmerlabs.scatterbrainsdk.HandshakeResult
import net.ballmerlabs.uscatterbrain.db.entities.DbMessage
import net.ballmerlabs.uscatterbrain.db.entities.HashlessScatterMessage
import net.ballmerlabs.uscatterbrain.network.BlockHeaderPacket
import net.ballmerlabs.uscatterbrain.network.BlockSequencePacket
import net.ballmerlabs.uscatterbrain.network.UpgradePacket
import net.ballmerlabs.uscatterbrain.network.bluetoothLE.BootstrapRequest
import java.io.File
import java.util.UUID
import java.util.concurrent.Flow

/**
 * dagger2 interface for WifiDirectRadioModule
 */
interface WifiDirectRadioModule {
    fun getBand(): Int

    fun addUke(uuid: UUID, bootstrap: UpgradePacket): Completable
    fun setUke(ukes: Map<UUID, UpgradePacket>): Completable
    fun removeUke(uuid: UUID)
    fun getUkes(): Map<UUID, UpgradePacket>

    fun bootstrapUke(band: Int, remoteLuid: UUID, selfLuid: UUID): Single<WifiDirectBootstrapRequest>
    fun bootstrapSeme(name: String, passphrase: String, band: Int, ownerPort: Int, self: UUID)

    fun safeShutdownGroup(): Completable

    fun awaitUke(): Observable<Pair<UUID, UpgradePacket>>

    fun getForceUke(): Boolean

    /**
     * Manually creates a wifi direct group, autogenerating the username/passphrase and
     * returning them as a BootstrapRequest
     * @return WifiDirectBootstrapRequest
     */
    fun createGroup(band: Int, remoteLuid: UUID, selfLuid: UUID): Flowable<WifiDirectBootstrapRequest>

    /**
     * Removes an existing wifi direct group if it exists
     * @return Completable
     */
    fun removeGroup(retries: Int = 10, delay: Int = 5): Completable

    /**
     * Ugly hack to determine if SoftAp or similar is hogging the wireless
     * adapter and blocking wifi direct
     *
     * @return Single emitting boolean, true if usable
     */
    fun wifiDirectIsUsable(): Single<Boolean>

    fun registerReceiver()

    fun unregisterReceiver()

    /**
     * Wrapper class combining BlockHeaderPacket, SequencePackets, and
     * a database entity
     *
     * Allows streaming messages from network directly into database/filestore
     *
     * @property headerPacket blockheader for this stream
     * @property sequencePacketsParam Flowable of sequence packets corresponding to the header packet
     * @property entity database entity for inserting this stream into the datastore, possibly autogenerated
     */
    class BlockDataStream(
        val headerPacket: BlockHeaderPacket,
        private val sequencePacketsParam: Flowable<BlockSequencePacket>,
        val entity: DbMessage?,
    ) {
        private val sequenceCompletable = CompletableSubject.create()
        val sequencePackets: Flowable<BlockSequencePacket> = sequencePacketsParam
            .doOnComplete { sequenceCompletable.onComplete() }
            .doOnError { e -> sequenceCompletable.onError(e) }

        constructor(
            headerPacket: BlockHeaderPacket,
            sequencePackets: Flowable<BlockSequencePacket>,
            filePath: File
        ) : this(
            entity = DbMessage.from(headerPacket, filePath),
            headerPacket = headerPacket,
            sequencePacketsParam = sequencePackets
        )

        fun await(): Completable {
            return sequenceCompletable
        }

        val toDisk: Boolean
            get() = headerPacket.isFile

        constructor(
            message: DbMessage,
            packetFlowable: Flowable<BlockSequencePacket>,
            todisk: Boolean = true
        ) : this(
            headerPacket = BlockHeaderPacket.newBuilder()
                .setToFingerprint(message.toFingerprint.firstOrNull())
                .setFromFingerprint(message.toFingerprint.firstOrNull())
                .setApplication(message.message.application)
                .setSig(message.message.sig)
                .setToDisk(todisk)
                .setSessionID(message.message.sessionid)
                .setMime(message.message.mimeType)
                .setExtension(message.message.extension)
                .setHashes(HashlessScatterMessage.hashes2hashProto(message.file.messageHashes))
                .setEndOfStream(false)
                .build(),
            entity = message,
            sequencePacketsParam = packetFlowable
        )

        companion object {
            fun endOfStream(): BlockDataStream {
                return BlockDataStream(
                    BlockHeaderPacket.newBuilder().setEndOfStream(true).build(),
                    Flowable.empty(),
                    File("/")
                )
            }
        }
    }
}